File: samples/javascript/data-integrity.js
/**
 * MITRE ATT&CK T1565: Data Manipulation - Integrity Verification
 * Educational demonstration of data integrity protection
 * 
 * @description Detects and prevents unauthorized data modifications
 * @reference https://attack.mitre.org/techniques/T1565/
 */

const crypto = require('crypto');

// =============================================================================
// DATA INTEGRITY MANAGER
// =============================================================================

/**
 * SECURE: Data integrity management system
 * Defense against ATT&CK T1565
 * 
 * @description Protects data integrity with HMAC signatures and audit trails
 * @security HMAC verification, audit trails, modification monitoring
 */
class DataIntegrityManager {
  constructor(secretKey) {
    this.secretKey = secretKey;
    
    // Audit trail: [{timestamp, user, recordId, field, oldValue, newValue, signature}, ...]
    this.auditTrail = [];
    
    // Change velocity tracking: user -> [{timestamp, recordId}, ...]
    this.userModifications = new Map();
    
    // Field sensitivity levels (higher = more sensitive)
    this.fieldSensitivity = {
      balance: 10,
      salary: 10,
      role: 9,
      permissions: 9,
      status: 8,
      email: 7,
      name: 5,
      description: 3
    };
    
    // Alert thresholds
    this.config = {
      massModificationThreshold: 10,
      massModificationWindowMs: 60 * 1000, // 60 seconds
      highValueChangeThreshold: 1000
    };
  }

  /**
   * Compute HMAC signature for a record
   * @param {string} recordId - Record identifier
   * @param {object} data - Record data
   * @returns {string} HMAC signature
   */
  computeSignature(recordId, data) {
    const dataCopy = {...data};
    delete dataCopy._signature;
    
    const canonical = JSON.stringify(dataCopy, Object.keys(dataCopy).sort());
    const message = `${recordId}:${canonical}`;
    
    return crypto.createHmac('sha256', this.secretKey)
      .update(message)
      .digest('hex');
  }

  /**
   * Sign a record with HMAC signature
   * @param {string} recordId - Record identifier
   * @param {object} data - Record data
   * @returns {object} Record with signature
   */
  signRecord(recordId, data) {
    const dataCopy = {...data};
    delete dataCopy._signature;
    
    const signature = this.computeSignature(recordId, dataCopy);
    
    return {
      ...dataCopy,
      _signature: signature
    };
  }

  /**
   * Verify record integrity
   * @param {string} recordId - Record identifier
   * @param {object} data - Record data
   * @returns {object} Verification result
   */
  verifyRecord(recordId, data) {
    if (!data._signature) {
      console.log(`❌ No signature found for record ${recordId}`);
      return {
        isValid: false,
        reason: 'No signature found - record may have been tampered with'
      };
    }

    const storedSignature = data._signature;
    const dataCopy = {...data};
    delete dataCopy._signature;
    
    const expectedSignature = this.computeSignature(recordId, dataCopy);
    
    const isValid = crypto.timingSafeEqual(
      Buffer.from(storedSignature),
      Buffer.from(expectedSignature)
    );

    if (!isValid) {
      console.log(`❌ Signature mismatch for record ${recordId}`);
      return {
        isValid: false,
        reason: 'Signature mismatch - record has been tampered with (T1565)'
      };
    }

    console.log(`✅ Record ${recordId} integrity verified`);
    return {
      isValid: true,
      reason: 'Record integrity verified'
    };
  }

  /**
   * Record a data change and check for anomalies
   * @param {object} changeDetails - Change details
   * @returns {object} Result with allowed flag, reason, and alerts
   */
  recordChange({user, recordId, field, oldValue, newValue}) {
    const now = Date.now();
    const alerts = [];

    // Create audit entry
    const auditEntry = {
      timestamp: now,
      user,
      recordId,
      field,
      oldValue,
      newValue
    };

    // Sign audit entry
    const auditSignature = crypto.createHmac('sha256', this.secretKey)
      .update(JSON.stringify(auditEntry))
      .digest('hex');

    this.auditTrail.push({
      ...auditEntry,
      signature: auditSignature
    });

    // Track modification for velocity analysis
    if (!this.userModifications.has(user)) {
      this.userModifications.set(user, []);
    }
    this.userModifications.get(user).push({timestamp: now, recordId});

    // Check 1: Mass modification detection
    const recentChanges = this.userModifications.get(user)
      .filter(m => now - m.timestamp <= this.config.massModificationWindowMs);

    if (recentChanges.length >= this.config.massModificationThreshold) {
      alerts.push(
        `ALERT (T1565): Mass modification detected! User ${user} modified ` +
        `${recentChanges.length} records in ${this.config.massModificationWindowMs / 1000}s`
      );
    }

    // Check 2: High-sensitivity field changes
    const fieldSensitivity = this.fieldSensitivity[field] || 1;
    if (fieldSensitivity >= 8) {
      alerts.push(
        `WARNING (T1565): High-sensitivity field '${field}' modified by ${user}`
      );
    }

    // Check 3: Suspicious value changes
    if (['balance', 'salary'].includes(field) && 
        typeof oldValue === 'number' && 
        typeof newValue === 'number') {
      const changeAmount = Math.abs(newValue - oldValue);
      if (changeAmount >= this.config.highValueChangeThreshold) {
        alerts.push(
          `ALERT (T1565): Large value change detected! ${field}: ` +
          `${oldValue} → ${newValue} (Δ${changeAmount})`
        );
      }
    }

    // Check 4: Multiple rapid changes to same record
    const recentRecordChanges = this.auditTrail
      .slice(-20)
      .filter(entry => 
        entry.recordId === recordId && 
        now - entry.timestamp <= 10000
      );

    if (recentRecordChanges.length >= 5) {
      alerts.push(
        `WARNING (T1565): Multiple rapid changes to record ${recordId}`
      );
    }

    // Determine if change should be allowed
    let allowed = true;
    let reason = 'Change recorded in audit trail';

    for (const alert of alerts) {
      if (alert.includes('ALERT') && alert.includes('Mass modification')) {
        allowed = false;
        reason = 'BLOCKED: Mass modification pattern detected (T1565)';
        break;
      }
    }

    if (alerts.length > 0) {
      console.log('⚠️  Anomalies detected:');
      alerts.forEach(alert => console.log(`   ${alert}`));
    }

    return {
      allowed,
      reason,
      alerts
    };
  }

  /**
   * Verify audit trail integrity
   * @returns {object} Verification result
   */
  verifyAuditTrail() {
    const issues = [];

    for (let i = 0; i < this.auditTrail.length; i++) {
      const entry = this.auditTrail[i];
      
      if (!entry.signature) {
        issues.push(`Entry ${i}: Missing signature`);
        continue;
      }

      const {signature, ...auditEntry} = entry;
      
      const expectedSignature = crypto.createHmac('sha256', this.secretKey)
        .update(JSON.stringify(auditEntry))
        .digest('hex');

      if (signature !== expectedSignature) {
        issues.push(
          `Entry ${i}: Audit trail tampering detected! (T1565) ` +
          `Record: ${entry.recordId}, User: ${entry.user}`
        );
      }
    }

    const isValid = issues.length === 0;

    if (isValid) {
      console.log('✅ Audit trail integrity verified');
    } else {
      console.log('❌ Audit trail tampering detected!');
      issues.forEach(issue => console.log(`   ${issue}`));
    }

    return {isValid, issues};
  }

  /**
   * Get change statistics
   * @param {string} user - Optional user filter
   * @returns {object} Statistics
   */
  getChangeReport(user = null) {
    const changes = user 
      ? this.auditTrail.filter(e => e.user === user)
      : this.auditTrail;

    const changesByUser = {};
    const changesByField = {};

    for (const entry of changes) {
      changesByUser[entry.user] = (changesByUser[entry.user] || 0) + 1;
      changesByField[entry.field] = (changesByField[entry.field] || 0) + 1;
    }

    const now = Date.now();
    const highVelocityUsers = [];
    
    for (const [username, modifications] of this.userModifications.entries()) {
      const recent = modifications.filter(
        m => now - m.timestamp <= 5 * 60 * 1000
      );
      if (recent.length >= 5) {
        highVelocityUsers.push({user: username, count: recent.length});
      }
    }

    return {
      totalChanges: changes.length,
      changesByUser,
      changesByField,
      highVelocityUsers: highVelocityUsers.sort((a, b) => b.count - a.count),
      auditTrailSize: this.auditTrail.length
    };
  }
}

// =============================================================================
// VULNERABLE IMPLEMENTATION (DO NOT USE IN PRODUCTION)
// =============================================================================

/**
 * VULNERABLE: Direct data modification with no integrity checks
 * 
 * Issues:
 * - No integrity verification
 * - No audit trail
 * - No change monitoring
 * - Attacker can modify data undetected
 * 
 * ATT&CK T1565: Attacker can tamper with data silently
 */
function vulnerableUpdateRecord(recordId, data) {
  // Just update the data - no verification or logging
  return data;
}

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

if (require.main === module) {
  console.log('='.repeat(80));
  console.log('MITRE ATT&CK T1565: Data Manipulation - Integrity Verification Demo');
  console.log('='.repeat(80));
  console.log('\nDetecting and preventing unauthorized data modifications\n');

  const secretKey = 'super_secret_integrity_key_DO_NOT_HARDCODE';
  const manager = new DataIntegrityManager(secretKey);

  // Scenario 1: Normal operations
  console.log('--- Scenario 1: Normal Operations ---\n');
  
  const recordId = 'user_12345';
  const userData = {
    name: 'Alice Smith',
    email: 'alice@example.com',
    balance: 1000,
    role: 'user'
  };

  const signedRecord = manager.signRecord(recordId, userData);
  console.log('Created and signed record:', recordId);

  manager.verifyRecord(recordId, signedRecord);

  manager.recordChange({
    user: 'admin',
    recordId,
    field: 'email',
    oldValue: 'alice@example.com',
    newValue: 'alice.smith@example.com'
  });

  // Scenario 2: Data tampering attempt
  console.log('\n--- Scenario 2: Data Tampering Attack (T1565) ---\n');
  
  const accountId = 'account_999';
  const accountData = {
    accountNumber: '999',
    balance: 5000,
    owner: 'Bob Jones'
  };

  const signedAccount = manager.signRecord(accountId, accountData);
  console.log('Original balance: $5000');

  const tamperedAccount = {...signedAccount};
  tamperedAccount.balance = 50000;
  console.log('Attacker changes balance to: $50000\n');

  manager.verifyRecord(accountId, tamperedAccount);

  // Scenario 3: Mass modification attack
  console.log('\n--- Scenario 3: Mass Modification Attack (T1565) ---\n');
  console.log('Attacker performing mass data modification...\n');
  
  const attacker = 'compromised_admin';
  
  for (let i = 0; i < 15; i++) {
    const result = manager.recordChange({
      user: attacker,
      recordId: `record_${i}`,
      field: 'status',
      oldValue: 'active',
      newValue: 'deleted'
    });

    if (!result.allowed) {
      console.log(`\n🛡️  Mass modification blocked after ${i + 1} attempts!`);
      console.log(`   Reason: ${result.reason}`);
      break;
    }
  }

  // Scenario 4: High-value field modification
  console.log('\n--- Scenario 4: High-Value Field Modification (T1565) ---\n');
  
  manager.recordChange({
    user: 'insider_threat',
    recordId: 'salary_database_001',
    field: 'salary',
    oldValue: 75000,
    newValue: 175000
  });

  // Scenario 5: Audit trail tampering
  console.log('\n--- Scenario 5: Audit Trail Tampering (T1565) ---\n');
  
  manager.recordChange({
    user: 'user1',
    recordId: 'rec1',
    field: 'status',
    oldValue: 'active',
    newValue: 'inactive'
  });

  console.log('Attacker attempts to modify audit trail...\n');
  
  if (manager.auditTrail.length > 0) {
    const original = manager.auditTrail[0];
    manager.auditTrail[0] = {
      ...original,
      user: 'attacker' // Tampered!
    };
  }

  manager.verifyAuditTrail();

  // Statistics
  console.log('\n--- Change Report ---\n');
  const report = manager.getChangeReport();
  console.log('Total changes:', report.totalChanges);
  console.log('Changes by user:', report.changesByUser);
  
  if (report.highVelocityUsers.length > 0) {
    console.log('\n⚠️  High-velocity users:');
    report.highVelocityUsers.forEach(({user, count}) => {
      console.log(`   ${user}: ${count} changes in 5 minutes`);
    });
  }

  console.log('\n' + '='.repeat(80));
  console.log('Protection mechanisms demonstrated:');
  console.log('  ✅ HMAC signatures for data integrity');
  console.log('  ✅ Tamper-evident audit trails');
  console.log('  ✅ Mass modification detection');
  console.log('  ✅ High-sensitivity field monitoring');
  console.log('  ✅ Audit trail integrity verification');
  console.log('  ✅ Modification velocity tracking');
  console.log('='.repeat(80));
}

module.exports = {
  DataIntegrityManager
};

File: samples/python/data_integrity.py
"""
MITRE ATT&CK T1565 - Data Manipulation: Integrity Verification

Data manipulation attacks involve unauthorized modification of data to compromise
integrity. This module demonstrates defensive measures to detect and prevent tampering.

Defenses implemented:
- HMAC-based record integrity verification
- Change audit trails with tamper-evident logging
- Modification velocity monitoring (detect mass changes)
- Field-level change tracking
- Cryptographic signatures for critical data

Educational purpose: Shows how to detect unauthorized data modifications and
maintain data integrity in the face of insider threats or compromised systems.
"""

import hmac
import hashlib
import json
import time
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple, Any
from collections import defaultdict


class DataIntegrityManager:
    """
    Manages data integrity using HMAC signatures and audit trails.
    
    Defends against T1565 (Data Manipulation) by:
    - Signing records with HMAC for integrity verification
    - Tracking all data modifications in tamper-evident audit log
    - Detecting unusual modification patterns
    - Monitoring field-level changes
    """
    
    def __init__(self, secret_key: str):
        """
        Initialize the integrity manager.
        
        Args:
            secret_key: Secret key for HMAC signing (keep secure!)
        """
        self.secret_key = secret_key.encode()
        
        # Audit trail: [(timestamp, user, record_id, field, old_value, new_value, signature), ...]
        self.audit_trail: List[Tuple[float, str, str, str, Any, Any, str]] = []
        
        # Change velocity tracking: {user: [(timestamp, record_id), ...]}
        self.user_modifications: Dict[str, List[Tuple[float, str]]] = defaultdict(list)
        
        # Field sensitivity levels (higher = more sensitive)
        self.field_sensitivity = {
            "balance": 10,
            "salary": 10,
            "role": 9,
            "permissions": 9,
            "status": 8,
            "email": 7,
            "name": 5,
            "description": 3
        }
        
        # Alert thresholds
        self.mass_modification_threshold = 10  # changes in short time
        self.mass_modification_window = 60  # seconds
        self.high_value_change_threshold = 1000.0
    
    def _compute_signature(self, record_id: str, data: Dict) -> str:
        """Compute HMAC signature for a record."""
        # Create canonical representation of data
        canonical = json.dumps(data, sort_keys=True)
        message = f"{record_id}:{canonical}".encode()
        
        signature = hmac.new(
            self.secret_key,
            message,
            hashlib.sha256
        ).hexdigest()
        
        return signature
    
    def sign_record(self, record_id: str, data: Dict) -> Dict:
        """
        Sign a record with HMAC signature for integrity verification.
        
        Args:
            record_id: Unique identifier for the record
            data: Record data to sign
        
        Returns:
            Record with '_signature' field added
        """
        data_copy = data.copy()
        
        # Remove signature field if present
        data_copy.pop('_signature', None)
        
        # Compute signature
        signature = self._compute_signature(record_id, data_copy)
        
        # Add signature to record
        data_copy['_signature'] = signature
        
        return data_copy
    
    def verify_record(self, record_id: str, data: Dict) -> Tuple[bool, Optional[str]]:
        """
        Verify the integrity of a signed record.
        
        Args:
            record_id: Unique identifier for the record
            data: Record data to verify
        
        Returns:
            Tuple of (is_valid, reason)
        """
        if '_signature' not in data:
            return False, "No signature found - record may have been tampered with"
        
        stored_signature = data['_signature']
        
        # Create copy without signature
        data_copy = data.copy()
        data_copy.pop('_signature', None)
        
        # Compute expected signature
        expected_signature = self._compute_signature(record_id, data_copy)
        
        # Constant-time comparison
        is_valid = hmac.compare_digest(stored_signature, expected_signature)
        
        if not is_valid:
            return False, "Signature mismatch - record has been tampered with (T1565)"
        
        return True, "Record integrity verified"
    
    def record_change(
        self,
        user: str,
        record_id: str,
        field: str,
        old_value: Any,
        new_value: Any
    ) -> Tuple[bool, str, List[str]]:
        """
        Record a data change in the audit trail and check for anomalies.
        
        Args:
            user: User making the change
            record_id: Record being modified
            field: Field being changed
            old_value: Previous value
            new_value: New value
        
        Returns:
            Tuple of (allowed, reason, alerts)
        """
        current_time = time.time()
        alerts = []
        
        # Create audit entry with signature
        audit_entry = (current_time, user, record_id, field, old_value, new_value)
        audit_signature = hmac.new(
            self.secret_key,
            json.dumps(audit_entry).encode(),
            hashlib.sha256
        ).hexdigest()
        
        # Store audit entry with signature
        self.audit_trail.append((*audit_entry, audit_signature))
        
        # Track modification for velocity analysis
        self.user_modifications[user].append((current_time, record_id))
        
        # Check 1: Mass modification detection
        recent_changes = [
            (ts, rid) for ts, rid in self.user_modifications[user]
            if current_time - ts <= self.mass_modification_window
        ]
        
        if len(recent_changes) >= self.mass_modification_threshold:
            alerts.append(
                f"ALERT (T1565): Mass modification detected! User {user} modified "
                f"{len(recent_changes)} records in {self.mass_modification_window}s"
            )
        
        # Check 2: High-value field changes
        field_sensitivity = self.field_sensitivity.get(field, 1)
        if field_sensitivity >= 8:
            alerts.append(
                f"WARNING (T1565): High-sensitivity field '{field}' modified by {user}"
            )
        
        # Check 3: Suspicious value changes (e.g., large balance increases)
        if field in ["balance", "salary"] and isinstance(old_value, (int, float)) and isinstance(new_value, (int, float)):
            change_amount = abs(new_value - old_value)
            if change_amount >= self.high_value_change_threshold:
                alerts.append(
                    f"ALERT (T1565): Large value change detected! {field}: "
                    f"{old_value} → {new_value} (Δ{change_amount})"
                )
        
        # Check 4: Multiple fields modified on same record (potential data corruption)
        recent_record_changes = [
            (ts, fld) for ts, user_id, rec_id, fld, _, _, _ in self.audit_trail[-20:]
            if rec_id == record_id and current_time - ts <= 10
        ]
        
        if len(recent_record_changes) >= 5:
            alerts.append(
                f"WARNING (T1565): Multiple rapid changes to record {record_id}"
            )
        
        # Determine if change should be allowed
        allowed = True
        reason = "Change recorded in audit trail"
        
        # Block if critical alert detected
        for alert in alerts:
            if "ALERT" in alert and "Mass modification" in alert:
                allowed = False
                reason = "BLOCKED: Mass modification pattern detected (T1565)"
                break
        
        return allowed, reason, alerts
    
    def verify_audit_trail(self) -> Tuple[bool, List[str]]:
        """
        Verify the integrity of the audit trail.
        
        Returns:
            Tuple of (is_valid, issues)
        """
        issues = []
        
        for i, entry in enumerate(self.audit_trail):
            if len(entry) != 7:
                issues.append(f"Entry {i}: Malformed audit entry")
                continue
            
            timestamp, user, record_id, field, old_val, new_val, signature = entry
            
            # Re-compute signature
            audit_entry = (timestamp, user, record_id, field, old_val, new_val)
            expected_signature = hmac.new(
                self.secret_key,
                json.dumps(audit_entry).encode(),
                hashlib.sha256
            ).hexdigest()
            
            if not hmac.compare_digest(signature, expected_signature):
                issues.append(
                    f"Entry {i}: Audit trail tampering detected! (T1565) "
                    f"Record: {record_id}, User: {user}"
                )
        
        is_valid = len(issues) == 0
        
        return is_valid, issues
    
    def get_change_report(self, user: Optional[str] = None) -> Dict:
        """Generate change statistics report."""
        current_time = time.time()
        
        if user:
            changes = [
                entry for entry in self.audit_trail
                if entry[1] == user
            ]
        else:
            changes = self.audit_trail
        
        # Count changes by user
        changes_by_user = defaultdict(int)
        for entry in changes:
            changes_by_user[entry[1]] += 1
        
        # Count changes by field
        changes_by_field = defaultdict(int)
        for entry in changes:
            changes_by_field[entry[3]] += 1
        
        # Find recent high-velocity users
        high_velocity_users = []
        for username, modifications in self.user_modifications.items():
            recent = [
                ts for ts, _ in modifications
                if current_time - ts <= 300  # Last 5 minutes
            ]
            if len(recent) >= 5:
                high_velocity_users.append((username, len(recent)))
        
        return {
            "total_changes": len(changes),
            "changes_by_user": dict(sorted(changes_by_user.items(), key=lambda x: x[1], reverse=True)),
            "changes_by_field": dict(sorted(changes_by_field.items(), key=lambda x: x[1], reverse=True)),
            "high_velocity_users": sorted(high_velocity_users, key=lambda x: x[1], reverse=True),
            "audit_trail_size": len(self.audit_trail)
        }


# =============================================================================
# VULNERABLE IMPLEMENTATION (DO NOT USE IN PRODUCTION)
# =============================================================================

def vulnerable_update_record(record_id: str, data: Dict) -> Dict:
    """
    VULNERABLE: Direct data modification with no integrity checks.
    
    Issues:
    - No integrity verification
    - No audit trail
    - No change monitoring
    - Attacker can modify data undetected
    
    ATT&CK T1565: Attacker can tamper with data silently.
    """
    # Just update the data - no verification or logging
    return data


# =============================================================================
# DEMONSTRATION
# =============================================================================

def simulate_normal_operations(manager: DataIntegrityManager) -> None:
    """Simulate normal data operations."""
    print("\n--- Normal Operations ---")
    
    # Create and sign a record
    record_id = "user_12345"
    user_data = {
        "name": "Alice Smith",
        "email": "alice@example.com",
        "balance": 1000.0,
        "role": "user"
    }
    
    signed_record = manager.sign_record(record_id, user_data)
    print(f"Created record: {record_id}")
    
    # Verify integrity
    is_valid, reason = manager.verify_record(record_id, signed_record)
    print(f"Integrity check: {reason}")
    
    # Record a normal change
    allowed, reason, alerts = manager.record_change(
        user="admin",
        record_id=record_id,
        field="email",
        old_value="alice@example.com",
        new_value="alice.smith@example.com"
    )
    print(f"Email update: {reason}")


def simulate_tampering_attempt(manager: DataIntegrityManager) -> None:
    """
    Simulate T1565 attack: Attacker tries to modify data without proper authorization.
    """
    print("\n--- Simulating Data Tampering Attack (T1565) ---")
    
    # Create legitimate record
    record_id = "account_999"
    account_data = {
        "account_number": "999",
        "balance": 5000.0,
        "owner": "Bob Jones"
    }
    
    signed_record = manager.sign_record(record_id, account_data)
    print(f"Original balance: ${account_data['balance']}")
    
    # Attacker tampers with balance
    tampered_record = signed_record.copy()
    tampered_record['balance'] = 50000.0  # Attacker adds a zero!
    
    print(f"Attacker changes balance to: ${tampered_record['balance']}")
    
    # Verify integrity (should fail)
    is_valid, reason = manager.verify_record(record_id, tampered_record)
    print(f"Integrity check result: {reason}")


def simulate_mass_modification_attack(manager: DataIntegrityManager) -> None:
    """
    Simulate T1565 attack: Attacker performs mass data modification.
    """
    print("\n--- Simulating Mass Modification Attack (T1565) ---")
    
    attacker = "compromised_admin"
    
    # Attacker tries to modify many records quickly
    for i in range(15):
        record_id = f"record_{i}"
        allowed, reason, alerts = manager.record_change(
            user=attacker,
            record_id=record_id,
            field="status",
            old_value="active",
            new_value="deleted"
        )
        
        if not allowed:
            print(f"  Attempt {i + 1}: BLOCKED!")
            print(f"    Reason: {reason}")
            break
        
        if alerts:
            for alert in alerts:
                print(f"    {alert}")
            if not allowed:
                break
        
        if i % 5 == 4:
            print(f"  Modified {i + 1} records so far...")


def simulate_high_value_fraud(manager: DataIntegrityManager) -> None:
    """
    Simulate T1565 attack: Attacker modifies high-value fields.
    """
    print("\n--- Simulating High-Value Field Modification (T1565) ---")
    
    allowed, reason, alerts = manager.record_change(
        user="insider_threat",
        record_id="salary_database_001",
        field="salary",
        old_value=75000.0,
        new_value=175000.0
    )
    
    print(f"Salary modification: {reason}")
    if alerts:
        for alert in alerts:
            print(f"  {alert}")


def simulate_audit_trail_tampering(manager: DataIntegrityManager) -> None:
    """
    Simulate T1565 attack: Attacker tries to modify audit trail.
    """
    print("\n--- Simulating Audit Trail Tampering (T1565) ---")
    
    # Add some legitimate entries
    manager.record_change("user1", "rec1", "status", "active", "inactive")
    manager.record_change("user2", "rec2", "balance", 100, 200)
    
    print("Audit trail created with 2 entries")
    
    # Attacker tries to tamper with audit trail
    if len(manager.audit_trail) > 0:
        print("Attacker attempts to modify audit entry...")
        original_entry = manager.audit_trail[0]
        
        # Tamper with the entry (change user)
        tampered_entry = (
            original_entry[0],  # timestamp
            "attacker",  # changed user
            original_entry[2],  # record_id
            original_entry[3],  # field
            original_entry[4],  # old_value
            original_entry[5],  # new_value
            original_entry[6]   # signature (now invalid)
        )
        
        manager.audit_trail[0] = tampered_entry
    
    # Verify audit trail
    is_valid, issues = manager.verify_audit_trail()
    
    if is_valid:
        print("  Audit trail is valid")
    else:
        print("  Audit trail tampering detected!")
        for issue in issues:
            print(f"    {issue}")


if __name__ == "__main__":
    print("=" * 70)
    print("MITRE ATT&CK T1565 - Data Manipulation: Integrity Verification")
    print("=" * 70)
    print("\nThis demo shows how to detect and prevent unauthorized data")
    print("modifications using cryptographic integrity verification and")
    print("behavioral monitoring.")
    
    # Create integrity manager with secret key
    secret_key = "super_secret_integrity_key_DO_NOT_HARDCODE"
    manager = DataIntegrityManager(secret_key)
    
    # Run simulations
    simulate_normal_operations(manager)
    simulate_tampering_attempt(manager)
    simulate_mass_modification_attack(manager)
    simulate_high_value_fraud(manager)
    simulate_audit_trail_tampering(manager)
    
    # Generate change report
    print("\n--- Change Report ---")
    report = manager.get_change_report()
    print(f"Total changes: {report['total_changes']}")
    print(f"Changes by user: {report['changes_by_user']}")
    
    if report['high_velocity_users']:
        print("\nHigh-velocity users (potential mass modification):")
        for user, count in report['high_velocity_users']:
            print(f"  {user}: {count} changes in 5 minutes")
    
    print("\n" + "=" * 70)
    print("Key Takeaways:")
    print("- Use HMAC signatures to ensure data integrity")
    print("- Maintain tamper-evident audit trails")
    print("- Monitor for mass modification patterns")
    print("- Track high-sensitivity field changes")
    print("- Verify audit trail integrity regularly")
    print("- Alert on unusual data modification patterns")
    print("=" * 70)
